### PR TITLE
[platform] Fix 'move to CPlatformXXX' todo

### PR DESCRIPTION
**Pull request by sy6sy2**
## Description

This PR tries to fix `//! @todo - move to CPlatformXXX` comments from `Application.cpp`.
The goal is to move platform specific initialization stuff in dedicated platform files.

## Motivation and Context

Keep a clean `Application.cpp` file (I guess).

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by sy6sy2** [comment]
@lrusak, before going further, can you check that I am in the right direction and if  I well understand the "todo"?
Thank you!

**Comment by lrusak** [comment]
> @lrusak, before going further, can you check that I am in the right direction and if I well understand the "todo"?
> Thank you!

tl;dr:
 - don't create more platform specifics (ask "how can I remove these platform ifdefs")
 - generalize the methods as much as possible (even if this means reordering).


long answer:

You have somewhat the right idea but the way you are doing it breaks the design. The whole reason to move to using `CPlatform` is to remove the platform specifics from `CApplication`. This includes the ifdeffery too.

`CPlatform` is constructed and init'd in `CServiceManager::InitStageOne`, this should be earlier enough so that `CPlatform` is alive for any of the calls you need to do.

Ideally we should just be able to put the platform specifics in the `CPlatformXXX::Init` method however this may not hold true and we may need to do things later in the init process. We also shouldn't have to access `CPlatform` through `CApplication` at all. If you need to call a method from `CPlatform` then you should add a method to `CServiceBroker` to get the member from `CServiceManager` (who owns the member) but it's likely that this isn't even needed at all.

**Comment by sy6sy2** [comment]
Thank you for taking time to explain me!
With your answer and Fuzzard's answer I understand better the main idea 😉 
I will try something.

**Comment by wsnipex** [comment]
Looks like this is the same as your xcode PR, branch mixup?

**Comment by sy6sy2** [comment]
Oups, force push on the wrong branch...

**Comment by wsnipex** [comment]
Ah ok, got it now, sorry for the confusion

**Comment by sy6sy2** [comment]
@lrusak can I modify signature of `Platform::Init()` to return a `bool` in order to handle error cases or it's a bad idea?
https://github.com/xbmc/xbmc/blob/cc286a397145184ff27458c610bfe839d5c432c7/xbmc/platform/Platform.h#L31-L36

Thank you!

**Comment by lrusak** [comment]
> @lrusak can I modify signature of `Platform::Init()` to return a `bool` in order to handle error cases or it's a bad idea?
> https://github.com/xbmc/xbmc/blob/cc286a397145184ff27458c610bfe839d5c432c7/xbmc/platform/Platform.h#L31-L36
> 
> Thank you!

sure if it makes sense to do so. Be aware you will have to change it for all platforms though.

**Comment by sy6sy2** [comment]
I hope my way is better now? 😇

File: xbmc/ServiceManager.cpp [hunk 1]
```diff
@@ -81,7 +81,10 @@ void CServiceManager::DeinitTesting()
 bool CServiceManager::InitStageOne()
 {
   m_Platform.reset(CPlatform::CreateInstance());
-  m_Platform->Init();
+  if (!m_Platform->Init())
+  {
```
Comment (lrusak) [inline xbmc/ServiceManager.cpp:85]: probably don't need the braces (same for all checks)

File: xbmc/Util.cpp [hunk 2]
```diff
@@ -2394,3 +2394,20 @@ int CUtil::GetRandomNumber()
 #endif
 }
 
+void CUtil::CopyUserDataIfNeeded(const std::string& strPath,
+                                 const std::string& file,
+                                 const std::string& destname)
+{
+  std::string destPath;
+  if (destname == "")
```
Comment (lrusak) [inline xbmc/Util.cpp:2402]: destname.empty()

File: xbmc/platform/overrides/darwin_embedded/PlatformDarwinEmbedded.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+#pragma once
+
+#include "platform/darwin/PlatformDarwin.h"
+
+class CPlatformDarwinEmbedded : public CPlatformDarwin
+{
+public:
+  /**\brief C'tor */
```
Comment (lrusak) [inline xbmc/platform/overrides/darwin_embedded/PlatformDarwinEmbedded.h:16]: doxygen doesn't provide any value here (you can remove it from all similar headers being added)

File: xbmc/platform/overrides/darwin_embedded/PlatformDarwinEmbedded.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  CPlatformDarwinEmbedded() = default;
+
+  /**\brief D'tor */
```
Comment (lrusak) [inline xbmc/platform/overrides/darwin_embedded/PlatformDarwinEmbedded.h:19]: same with this

File: xbmc/platform/overrides/windows/PlatformWin32.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+{
+  if (!CPlatform::Init())
+  {
+    return false;
+  }
+
+  CEnvironment::setenv("OS", "win32"); // for python scripts that check the OS
+
+  CWIN32Util::SetThreadLocalLocale(
+      true); // enable independent locale for each thread, see https://connect.microsoft.com/VisualStudio/feedback/details/794122
```
Comment (lrusak) [inline xbmc/platform/overrides/windows/PlatformWin32.cpp:29]: ugly line break (just move the comment)

File: xbmc/platform/overrides/windowsstore/PlatformWin10.cpp [hunk 6]
```diff
@@ -18,11 +19,17 @@ CPlatform* CPlatform::CreateInstance()
...
 {
+  if (!CPlatform::Init())
+  {
+    return false;
+  }
   CEnvironment::setenv("SSL_CERT_FILE", CSpecialProtocol::TranslatePath("special://xbmc/system/certs/cacert.pem").c_str(), 1);
+  CEnvironment::setenv("OS", "win32"); // for python scripts that check the OS
+
+  CWIN32Util::SetThreadLocalLocale(
+      true); // enable independent locale for each thread, see https://connect.microsoft.com/VisualStudio/feedback/details/794122
```
Comment (lrusak) [inline xbmc/platform/overrides/windowsstore/PlatformWin10.cpp:32]: same thing here

**Review by lrusak — CHANGES_REQUESTED** [review]
Looks much better! I've just pointed out some minors.

Also, please use the same copyright header for all new files being added (I think it should be 2005-2020 ???)

**Comment by fuzzard** [comment]
Did a runtime test of this on Android-arm64, seems fine. 
@KarellenX  did a test on win64, and said it seemed ok.

Any linux folks able to test it on that front?

**Review by fuzzard — APPROVED** [review]

**Comment by sy6sy2** [comment]
I did the rebase, I hope I did not break anything concerning what @lrusak did on #18534 😕

**Comment by sy6sy2** [comment]
Jenkins build this please

**Review by lrusak — APPROVED** [review]
Looks good. Great work!

I'll likely submit a PR after this to clean up some stuff and remove the weird /overrides directory. We shouldn't really be using wildcards to add source files to the cmake build.
